### DISCUSSION TITLE
Will this library work with NativeScript

### DISCUSSION THREAD

**Discussion by leoncc** [cyrilletuzi/angular-async-local-storage#11] [Q&A] [answered]
Do you foresee any issues using angular-async-local-storage with NativeScript?

**Comment by cyrilletuzi** [comment]
It's just a normal Angular library, no extra dependencies, so yes it should work with NativeScript.

**Comment by cyrilletuzi** [comment]
Of course, IndexedDB API is required for this API to work. I'm not sure it's the case in NativeScript., as it doesn't produce a web app but a native app.

**Comment by leoncc** [comment]
Thanks. The IndexedDB was what I was unsure of as it seems to be available only in browsers.

**Comment by cyrilletuzi** [comment, marked as answer]
For everyone : after deeper investigation, this lib won't work in NativeScript as NS doesn't support local storage browsers API (neither `localStorage` or `IndexedDB`).

You'll have to use the [specific APIs of NativeScript](https://stackoverflow.com/questions/39463157/is-there-localstorage-in-nativescript).

**Comment by schlusslicht** [comment]
This seems to be possible (untested!). One can inject a facade for the [NativeScript ApplicationSettings](https://docs.nativescript.org/ns-framework-modules/application-settings) into the [StorageMap](https://github.com/cyrilletuzi/angular-async-local-storage/blob/master/projects/ngx-pwa/local-storage/src/lib/storages/storage-map.service.ts) service, when running in a native environment.

`native.module.tns.ts`
```
@NgModule({
  providers: [
    { provide: LocalDatabase, useClass: NativeStorageDatabase }
  ]
})

export class NativeModule { }
```

`native.database.tns.ts`
```
import { Injectable } from '@angular/core';
import { LocalDatabase, StorageMap } from '@ngx-pwa/local-storage';
import { asyncScheduler, Observable, of, throwError } from 'rxjs';
import { observeOn } from 'rxjs/operators';
import * as nativeStorage from 'tns-core-modules/application-settings';

@Injectable({
  providedIn: StorageMap
})

export class NativeStorageDatabase implements LocalDatabase {

  public get size(): Observable<number> {
    return of(nativeStorage.getAllKeys().length);
  }

  public clear(): Observable<undefined> {
    nativeStorage.clear();
    return of(undefined);
  }

  public delete(key: string): Observable<undefined> {
    nativeStorage.remove(key);
    return of(undefined);
  }

  public get<T = any>(key: string): Observable<T | undefined> {
    const data = nativeStorage.getString(key);

    if (data === null || data === undefined) {
      return of(undefined);
    }

    try {
      return of(JSON.parse(data) as T);
    } catch (error) {
      return throwError(error as SyntaxError);
    }
  }

  public has(key: string): Observable<boolean> {
    return nativeStorage.hasKey(key) ? of(true) : of(false);
  }

  public keys(): Observable<string> {
    return new Observable<string>((observer) => {
      nativeStorage.getAllKeys().forEach((key) => observer.next(key));
      observer.complete();
    }).pipe(observeOn(asyncScheduler));
  }

  public set(key: string, data: any): Observable<undefined> {
    let json: string | null = null;

    try {
      json = JSON.stringify(data);
    } catch (error) {
      return throwError(error as TypeError);
    }

    try {
      nativeStorage.setString(key, json);
      return of(undefined);
    } catch (error) {
      return throwError(error as DOMException);
    }
  }

}
```

**Comment by cyrilletuzi** [comment]
@schlusslicht Thanks for your interest in this lib and for this example. It should indeed be possible for the user to do its own implementation.
